overlord/servicestate/servicemgr.go

```diff
@@ -42,8 +54,224 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
+// on the system before or during the unit's transition to inactive.
+// XXX: move to systemd package ?
+func inactiveEnterTimestampForUnit(unit string) (time.Time, error) {
+	// XXX: ignore stderr of systemctl command to avoid further infractions
+	//      around LP #1885597
+	out, err := exec.Command("systemctl", "show", "--property", "InactiveEnterTimestamp", unit).Output()
+	if err != nil {
+		return time.Time{}, osutil.OutputErr(out, err)
+	}
+
+	// the time returned by systemctl here will be formatted like so:
+	// InactiveEnterTimestamp=Fri 2021-04-16 15:32:21 UTC
+	// so we have to parse the time with a matching Go time format
+	splitVal := strings.SplitN(string(out), "=", 2)
+	if len(splitVal) != 2 {
+		// then we don't have an equals sign in the output, so systemctl must be
+		// broken
+		return time.Time{}, fmt.Errorf("internal error: systemctl output (%s) is malformed", string(out))
+	}
+
+	if splitVal[1] == "" {
+
+		return time.Time{}, nil
+	}
+
+	// finally parse the time string
+	inactiveEnterTime, err := time.Parse("Mon 2006-01-02 15:04:05 MST", splitVal[1])
+	if err != nil {
+		return time.Time{}, fmt.Errorf("internal error: systemctl time output (%s) is malformed", splitVal[1])
+	}
+	return inactiveEnterTime, nil
+}
+
+func (m *ServiceManager) ensureSnapServicesUpdated() (err error) {
+	m.state.Lock()
+	defer m.state.Unlock()
+	if m.ensuredSnapSvcs {
+		return nil
+	}
+
+	// only run after we are seeded
+	var seeded bool
+	err = m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
+
+	// we are seeded, now we need to find all snap services and re-generate
+	// services as necessary
+
+	// ensure all snap services are updated
+	allStates, err := snapstate.All(m.state)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	snapsMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+
+	for _, snapSt := range allStates {
+		info, err := snapSt.CurrentInfo()
+		if err != nil {
+			return err
+		}
+
+		// TODO: handle vitality rank here too from configcore
+		snapsMap[info] = nil
+	}
+
+	// setup ensure options
+	ensureOpts := &wrappers.EnsureSnapServicesOptions{
+		Preseeding: snapdenv.Preseeding(),
+	}
+
+	// set RequireMountedSnapdSnap if we are on UC18+ only
+	deviceCtx, err := devicestate.DeviceCtx(m.state, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
+		ensureOpts.RequireMountedSnapdSnap = true
+	}
+
+	// TODO: should we use an actual interacter here ?
+	modified, err := wrappers.EnsureSnapServices(snapsMap, ensureOpts, progress.Null)
+	if err != nil {
+		return err
+	}
+
+	// if nothing was modified or we are not on UC18+, we are done
+	if len(modified) == 0 || deviceCtx.Classic() || deviceCtx.Model().Base() == "" {
+		m.ensuredSnapSvcs = true
+		return nil
+	}
+
+	// otherwise we need to check for all the services that were modified if
```

Comment:
the code below here until setting the flag excluded should live in its own helper and possible file named after the bug or something

Reply:
I meant with below  162-264 and the helper it uses should live separately

Reply:
this is done